src/mctpd.c

```diff
@@ -999,6 +1012,83 @@ static int handle_control_get_message_type_support(
 	return rc;
 }
 
+static int
+handle_control_get_vdm_type_support(struct ctx *ctx, int sd,
+				    const struct sockaddr_mctp_ext *addr,
+				    const uint8_t *buf, const size_t buf_size)
+{
+	size_t resp_len, max_rsp_len, vdm_count, vid_size_left;
+	struct mctp_ctrl_resp_get_vdm_support *resp = NULL;
+	struct mctp_ctrl_cmd_get_vdm_support *req = NULL;
+	struct vdm_type_support *cur_vdm;
+	uint8_t *resp_buf;
+	int rc;
+
+	if (buf_size < sizeof(*req)) {
+		warnx("short Get Message Type Support message");
+		return -ENOMSG;
+	}
+
+	req = (void *)buf;
+	vdm_count = ctx->num_supported_vdm_types;
+	// Allocate space for 32 bit VID + 16 bit cmd set
+	max_rsp_len = sizeof(*resp) + sizeof(uint16_t);
+	resp_buf = malloc(max_rsp_len);
+	if (!resp_buf) {
+		warnx("Failed to allocate response buffer");
+		return -ENOMEM;
+	}
+	resp = (void *)resp_buf;
+	mctp_ctrl_msg_hdr_init_resp(&resp->ctrl_hdr, req->ctrl_hdr);
+
+	if (vdm_count == 0) {
+		if (ctx->verbose) {
+			warnx("Get VDM Type Support but no VDMs registered");
+		}
+		resp_len = sizeof(struct mctp_ctrl_resp);
+		resp->completion_code = MCTP_CTRL_CC_ERROR;
+	} else if (req->vendor_id_set_selector >= vdm_count) {
+		if (ctx->verbose) {
+			warnx("Get VDM Type Support selector %u out of range (max %zu)",
+			      req->vendor_id_set_selector, vdm_count);
+		}
+		resp_len = sizeof(struct mctp_ctrl_resp);
+		resp->completion_code = MCTP_CTRL_CC_ERROR_INVALID_DATA;
+	} else {
+		cur_vdm =
+			&ctx->supported_vdm_types[req->vendor_id_set_selector];
+		vid_size_left = (cur_vdm->vid_format ==
+				 MCTP_GET_VDM_SUPPORT_PCIE_FORMAT_ID) ?
+					sizeof(uint16_t) :
+					0;
+		resp_len = max_rsp_len - vid_size_left;
+
+		resp->completion_code = MCTP_CTRL_CC_SUCCESS;
+		resp->vendor_id_set_selector =
+			req->vendor_id_set_selector == (vdm_count - 1) ?
+				MCTP_GET_VDM_SUPPORT_NO_MORE_CAP_SET :
+				req->vendor_id_set_selector + 1;
+		resp->vendor_id_format = cur_vdm->vid_format;
+		uint8_t *out_ptr = (uint8_t *)(resp + 1);
+		out_ptr -= vid_size_left;
+
+		if (cur_vdm->vid_format ==
+		    MCTP_GET_VDM_SUPPORT_PCIE_FORMAT_ID) {
+			resp->vendor_id_data_pcie =
+				htobe16(cur_vdm->vendor_id.pcie);
+		} else {
+			resp->vendor_id_data_iana =
+				htobe32(cur_vdm->vendor_id.iana);
+		}
```

Comment:
There are three separate conditionals that depend on `->vid_format` here. perhaps you can simplify these a little?